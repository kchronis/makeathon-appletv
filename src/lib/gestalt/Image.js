/**
 * WIP - Gestalt like Image
 * https://github.com/pinterest/gestalt/tree/master/src/Image
 * @flow
 */

import * as React from 'react';
import { View, Image as ReactImage } from 'react-native';
import type { StyleObj } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import Text from './Text';
import RemoteImageView from '../../lib/react-native-remote-image-view';

import PIColor from '../styles/picolor';

const resolveAssetSource = require('resolveAssetSource');

type Props = {|
  // alt: string,
  children?: React.Node,
  color: string,
  // fit: 'contain' | 'cover' | 'none', // Not supported in React Native
  // naturalHeight: number,             // Not supported in React Native
  // naturalWidth: number,              // Not supported in React Native
  local: boolean, // Only available in React Native
  onError?: () => void,
  onLoad?: () => void,
  // sizes?: string,                    // Not supported in React Native
  src: string,
  // srcSet?: string,                   // Not supported in React Native

  style?: StyleObj, // React Native only: Should go away eventually
|};

export default class Image extends React.PureComponent<Props> {
  static defaultProps = {
    color: 'transparent',
    local: false,
  };
  render() {
    const { color, children, local, onLoad, onError, src, style } = this.props;
    const imageStyle = [
      {
        // Let the container component define the width and height by default
        width: '100%',
        height: '100%',
      },
      style,
    ];
    const childContent = children ? (
      <View
        style={{
          position: 'absolute',
          top: 0,
          left: 0,
          bottom: 0,
          right: 0,
          overflow: 'hidden',
        }}
      >
        {children}
      </View>
    ) : null;
    return (
      <View
        style={[
          {
            position: 'relative',
            backgroundColor: color,
          },
        ]}
      >
        {local ? (
          <ReactImage
            style={imageStyle}
            source={src}
            onLoad={onLoad}
            onError={onError}
          />
        ) : (
          <RemoteImageView
            style={imageStyle}
            imageURL={src}
            onLoad={onLoad}
            onError={onError}
          />
        )}

        {childContent}
      </View>
    );
  }
}

// Playground support

// eslint-disable-next-line import/first
import { type PlaygroundDefineType } from '../../screens/Playground/Playground';

module.exports.__cards__ = (define: PlaygroundDefineType) => {
  define('Load Remote Image', () => (
    <View
      style={{
        backgroundColor: PIColor.primaryBackgroundColor,
        flex: 1,
        justifyContent: 'center',
        padding: 10,
      }}
    >
      <Text style={{ paddingBottom: 5 }}>Remote Image:</Text>
      <Image
        src={
          'https://i.pinimg.com/736x/92/fd/59/92fd59efefc446c7e858ebd16c53feba--pinterest-logo-png-pinterest-board.jpg'
        }
        // A size needs to be given or the container needs to provide one
        style={{ width: 100, height: 100, alignSelf: 'center' }}
      />
    </View>
  ));

  define('Local Image', () => {
    // Example to resolve the size of a local image
    // eslint-disable-next-line global-require
    const localImage = require('../../screens/Playground/images/pinterest_logo.jpg');
    const localImageSource = resolveAssetSource(localImage);
    return (
      <View
        style={{
          backgroundColor: PIColor.primaryBackgroundColor,
          flex: 1,
          justifyContent: 'center',
          padding: 10,
        }}
      >
        <Text style={{ paddingBottom: 5 }}>Local Image:</Text>
        {/* $FlowFixMe: Until flow 0.53 exact prop types in same
    file is not supported */}
        <Image
          local
          src={localImage}
          style={{
            width: localImageSource ? localImageSource.width : 100,
            height: localImageSource ? localImageSource.height : 100,
            alignSelf: 'center',
          }}
        />
      </View>
    );
  });
};
