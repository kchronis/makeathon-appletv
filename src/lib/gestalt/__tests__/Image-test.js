/* eslint-env jest */

import React from 'react';
import { View } from 'react-native';
import renderer from 'react-test-renderer';

import Image from '../Image';

test('Image matches snapshot', () => {
  const component = renderer.create(<Image src="foo.png" />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Image local matches snapshot', () => {
  /* eslint-disable global-require */
  const component = renderer.create(
    <View>
      <Image local src={require('./img/img1.png')} />
      <Image local src={require('./img/img2.png')} />
    </View>
  );
  /* eslint-enable global-require */
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
