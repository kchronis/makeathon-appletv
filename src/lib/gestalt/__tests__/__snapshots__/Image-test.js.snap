// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image local matches snapshot 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "position": "relative",
        },
      ]
    }
  >
    <Image
      onError={undefined}
      onLoad={undefined}
      source={
        Object {
          "testUri": "../../../src/lib/gestalt/__tests__/img/img1.png",
        }
      }
      style={
        Array [
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "position": "relative",
        },
      ]
    }
  >
    <Image
      onError={undefined}
      onLoad={undefined}
      source={
        Object {
          "testUri": "../../../src/lib/gestalt/__tests__/img/img2.png",
        }
      }
      style={
        Array [
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`Image matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "position": "relative",
      },
    ]
  }
>
  <MSRRemoteImageView
    imageURL={null}
    onError={undefined}
    onLayout={[Function]}
    onLoad={undefined}
    placeholderBackgroundColor={4292072403}
    style={
      Array [
        Array [
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ],
        null,
      ]
    }
  />
</View>
`;
